Fix docker images build

colin-tim opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #
Fix docker images build #1778
## 🗣 Description ##
updated base images for backed, worker, docs and frontend for consistency (all use node 18) and in order to resolve a lot of vulnerabilities in the previous used images. 
updated the version for python in worker dockerfile. 
moved the docs to use a dockerfile in the parent folder (see docker-compose.yml). 
updated dependencies. 

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##
Followed the steps in Quick start guide (https://docs.crossfeed.cyber.dhs.gov/dev/quickstart/#initial-setup) and everything built and ran as expected. Able to see the frontend and docs site.


## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [ ] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [ ] Tests have been added and/or modified to cover the changes in this PR.
- [ ] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Create a release.

mbachtell-nais commented:
My local testing got stuck at building pandas in Python.  We will talk at the 10:00 scrum.
![image](https://user-images.githubusercontent.com/126006749/230895419-3c35a439-6d9e-49df-8426-4347fedb5aae.png)

epicfaace commented on docker-compose.yml:
```diff
@@ -59,7 +59,7 @@ services:
     build:
       context: ./
-      dockerfile: ./docs/Dockerfile
+      dockerfile: ./Dockerfile.docs
```
Why change this path?

colin-tim commented:
Because the context is the parent, it failed to build while building the
docs service, if you want, you can change it to point back to the original,
which I updated, but you will see that it fails

On Mon, Apr 10, 2023 at 6:54 AM Ashwin Ramaswami ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In docker-compose.yml
> <https://github.com/cisagov/crossfeed/pull/1792#discussion_r1161739576>:
>
> > @@ -59,7 +59,7 @@ services:
>    docs:
>      build:
>        context: ./
> -      dockerfile: ./docs/Dockerfile
> +      dockerfile: ./Dockerfile.docs
>
> Why change this path?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cisagov/crossfeed/pull/1792#pullrequestreview-1377563887>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATVZG4YQ257DFVD5N5SC2RDXAQGITANCNFSM6AAAAAAWYQYRPY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

epicfaace commented on backend/Dockerfile.worker:
```diff
@@ -27,11 +27,11 @@ RUN go mod init crossfeed-worker
 WORKDIR /app
 
-RUN apk add --update --no-cache wget build-base curl git unzip openssl-dev linux-headers python3=3.10.10-r0 python3-dev py3-pip ruby ruby-dev zlib-dev libffi-dev libxml2-dev libxslt-dev postgresql-dev gcc musl-dev py3-pandas py3-scikit-learn cargo
+RUN apk add --update --no-cache wget build-base curl git unzip openssl-dev linux-headers python3=3.10.11-r0 python3-dev py3-pip ruby ruby-dev zlib-dev libffi-dev libxml2-dev libxslt-dev postgresql-dev gcc musl-dev py3-pandas py3-scikit-learn cargo
```
Is it possible that it's failing because many packages do not have wheels already available for Python 3.10.11? (and maybe better to downgrade to a version of python where these wheels are already available?) just thinking...

colin-tim replied:
3.10.10 wasn’t in the image, we have to use the latest image otherwise the
build will fail.

colin-tim replied:
just added scikit-learn==1.2.2 to work around old package.

colin-tim commented:
3.10.10 wasn’t in the image, we have to use the latest image otherwise the
build will fail.
did you try running pip install -r worker/requirements.txt first?

On Mon, Apr 10, 2023 at 6:57 AM Ashwin Ramaswami ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In backend/Dockerfile.worker
> <https://github.com/cisagov/crossfeed/pull/1792#discussion_r1161742186>:
>
> >
>  WORKDIR /app
>
> -RUN apk add --update --no-cache wget build-base curl git unzip openssl-dev linux-headers python3=3.10.10-r0 python3-dev py3-pip ruby ruby-dev zlib-dev libffi-dev libxml2-dev libxslt-dev postgresql-dev gcc musl-dev py3-pandas py3-scikit-learn cargo
> +RUN apk add --update --no-cache wget build-base curl git unzip openssl-dev linux-headers python3=3.10.11-r0 python3-dev py3-pip ruby ruby-dev zlib-dev libffi-dev libxml2-dev libxslt-dev postgresql-dev gcc musl-dev py3-pandas py3-scikit-learn cargo
>
> Is it possible that it's failing because many packages do not have wheels
> already available for Python 3.10.11? (and maybe better to downgrade to a
> version of python where these wheels are already available?) just
> thinking...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cisagov/crossfeed/pull/1792#pullrequestreview-1377567993>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATVZG47BXJAMBSQO5LQQFYTXAQGUDANCNFSM6AAAAAAWYQYRPY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

epicfaace commented:
> Because the context is the parent, it failed to build while building the docs service, if you want, you can change it to point back to the original, which I updated, but you will see that it fails
> […](#)
> On Mon, Apr 10, 2023 at 6:54 AM Ashwin Ramaswami ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In docker-compose.yml <[#1792 (comment)](https://github.com/cisagov/crossfeed/pull/1792#discussion_r1161739576)>: > @@ -59,7 +59,7 @@ services: docs: build: context: ./ - dockerfile: ./docs/Dockerfile + dockerfile: ./Dockerfile.docs Why change this path? — Reply to this email directly, view it on GitHub <[#1792 (review)](https://github.com/cisagov/crossfeed/pull/1792#pullrequestreview-1377563887)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ATVZG4YQ257DFVD5N5SC2RDXAQGITANCNFSM6AAAAAAWYQYRPY> . You are receiving this because you authored the thread.Message ID: ***@***.***>

Ok, then in that case we should delete `docs/Dockerfile`.

colin-tim commented:
> > Because the context is the parent, it failed to build while building the docs service, if you want, you can change it to point back to the original, which I updated, but you will see that it fails
> > […](#)
> > On Mon, Apr 10, 2023 at 6:54 AM Ashwin Ramaswami _**@**_._**> wrote: _**@**_.**_* commented on this pull request. ------------------------------ In docker-compose.yml <[#1792 (comment)](https://github.com/cisagov/crossfeed/pull/1792#discussion_r1161739576)>: > @@ -59,7 +59,7 @@ services: docs: build: context: ./ - dockerfile: ./docs/Dockerfile + dockerfile: ./Dockerfile.docs Why change this path? — Reply to this email directly, view it on GitHub <[#1792 (review)](https://github.com/cisagov/crossfeed/pull/1792#pullrequestreview-1377563887)>, or unsubscribe https://github.com/notifications/unsubscribe-auth/ATVZG4YQ257DFVD5N5SC2RDXAQGITANCNFSM6AAAAAAWYQYRPY . You are receiving this because you authored the thread.Message ID: _**@**_.***>
> 
> Ok, then in that case we should delete `docs/Dockerfile`.

removed the old docs dockerfile as per comment.

epicfaace commented on backend/Dockerfile.worker:
```diff
@@ -41,7 +41,11 @@ RUN gem install bundler:2.3.21
 # Install pe-source module
+# why we sync to a specific commit in cf-staging branch?
+# consider syncing to latest in a stable branch do avoid churn
 RUN git clone https://github.com/cisagov/pe-reports.git && cd pe-reports && git checkout c9cbbd73b22ef38cabe1da6ba50aeb2dc0be4f99 && pip install . 
```
@aloftus23

aloftus23 replied:
P&E set up a dedicated branch to be installed in Crossfeed's worker container because we only want to install necessary packages and it's setup to grab variables from the parameter store. 

The main pe-reports branch has a few functionalities including reporting, mailing, and data collection; we only want the data collection portion.  

I suggested a change that will make sure it's the latest.
```suggestion
RUN git clone -b cf-source-staging https://github.com/cisagov/pe-reports.git && cd pe-reports && pip install .
```

colin-tim replied:
Updated as per suggestion.

mbachtell-nais reviewed (commented):
Everything built locally for me but the docs and mariadb containers. These did not start. 
![image](https://user-images.githubusercontent.com/126006749/231743877-2710cdd7-7a0e-4d78-b1b8-cf824e9cc961.png)

The elasticsearch container is throwing a flood warning but I think that might be a local/me issue rather than a build issue.

epicfaace commented:
@colin-tim  localhost:4000 (docs site) looks like this for me. It should look like https://docs.crossfeed.cyber.dhs.gov/. Can you check and fix?

<img width="884" alt="image" src="https://user-images.githubusercontent.com/1689183/232067753-8a2c008f-771d-4f2b-a8cc-5bbbb2e0e315.png">

colin-tim commented:
Fixed the issues raised by @epicfaace. See my latest push. 
Waiting for approval to merge to master.

epicfaace reviewed (commented):
@colin-tim thanks for fixing the docs!

Now, when I go to `http://localhost` and log in to Crossfeed, it just shows me a blank screen:

<img width="1381" alt="image" src="https://user-images.githubusercontent.com/1689183/232181249-5e5b2ce9-ec2d-46a3-8747-7a1d7aeb0881.png">

js console errors:

<img width="1453" alt="image" src="https://user-images.githubusercontent.com/1689183/232181269-448ab394-c2ce-488d-897e-52c9228996c0.png">


errors on backend endpoint:

<img width="640" alt="image" src="https://user-images.githubusercontent.com/1689183/232181267-fe665dd0-a3b6-4191-9118-5af4e4a9b66f.png">

mbachtell-nais commented:
> 

I'm seeing this same callback response on other builds too.

epicfaace commented:
> > 
> 
> I'm seeing this same callback response on other builds too.

Do you mean that you're seeing it on the master branch / other branches?

mbachtell-nais commented:
> 

I cloned a new version this morning and had to make the changes that were suggested here in backend/Dockerfile.worker.  We should probably talk tomorrow about this.

[Johnng007](https://github.com/Johnng007) commented [3 days ago](https://github.com/cisagov/crossfeed/issues/1778#issuecomment-1507271459)
updating Python 3 library to 3.10.11 r0 in backend Dockerfile.worker
and also updating FROM node:16.19.0-alpine3.17 as build to FROM node:16-alpine3.14 as build
worked for me. Now its Building.
